test/gtest/uct/test_p2p_mix.cc

```diff
@@ -0,0 +1,217 @@
+    }
+
+    ucs_status_t put_bcopy(const mapped_buffer &sendbuf,
+                           const mapped_buffer &recvbuf,
+                           uct_completion_t *comp)
+    {
+        ssize_t packed_len;
+        packed_len = uct_ep_put_bcopy(sender().ep(0), mapped_buffer::pack,
+                                      (void*)&sendbuf, recvbuf.addr(), recvbuf.rkey());
+        if (packed_len >= 0) {
+            EXPECT_EQ(sendbuf.length(), (size_t)packed_len);
+            return UCS_OK;
+        } else {
+            return (ucs_status_t)packed_len;
+        }
+    }
+
+    ucs_status_t am_short(const mapped_buffer &sendbuf,
+                          const mapped_buffer &recvbuf,
+                          uct_completion_t *comp)
+    {
+        ucs_status_t status;
+        status = uct_ep_am_short(sender().ep(0), AM_ID,
+                                 *(uint64_t*)sendbuf.ptr(),
+                                 (uint64_t*)sendbuf.ptr() + 1,
+                                 sendbuf.length() - sizeof(uint64_t));
+        if (status == UCS_OK) {
+            ucs_atomic_add32(&am_pending, +1);
+        }
+        return status;
+    }
+
+    ucs_status_t am_zcopy(const mapped_buffer &sendbuf,
+                          const mapped_buffer &recvbuf,
+                          uct_completion_t *comp)
+    {
+        ucs_status_t status;
+        uct_iov_t iov;
+
+        iov.buffer = sendbuf.ptr();
+        iov.count  = 1;
+        iov.length = sendbuf.length();
+        iov.memh   = sendbuf.memh();
+        status = uct_ep_am_zcopy(sender().ep(0), AM_ID, NULL, 0, &iov, 1, comp);
+        if (status == UCS_OK || status == UCS_INPROGRESS) {
+            ucs_atomic_add32(&am_pending, +1);
+        }
+        return status;
+    }
+
+    void random_op(const mapped_buffer &sendbuf, const mapped_buffer &recvbuf)
+    {
+        uct_completion_t comp;
+        ucs_status_t status;
+        int op;
+
+        op         = rand() % m_avail_send_funcs.size();
+        comp.count = 1;
+        comp.func  = completion_callback;
+
+        for (;;) {
+            status = (this->*m_avail_send_funcs[op])(sendbuf, recvbuf, &comp);
+            if (status == UCS_OK) {
+                break;
+            } else if (status == UCS_INPROGRESS) {
+                /* coverity[loop_condition] */
+                while (comp.count > 0) {
+                    progress();
+                }
+                break;
+            } else if (status == UCS_ERR_NO_RESOURCE) {
+                progress();
+                continue;
+            } else {
+                ASSERT_UCS_OK(status);
+            }
+        }
+    }
+
+protected:
+    virtual void init() {
+        uct_p2p_test::init();
+        ucs_status_t status = uct_iface_set_am_handler(receiver().iface(), AM_ID,
+                                                       am_callback, NULL,
+                                                       UCT_AM_CB_FLAG_ASYNC);
+        ASSERT_UCS_OK(status);
+
```

Comment:
I would propose to add `m_avail_send_funcs.reserve(10);` to make this routine slightly faster.